bcm27xx: add support for Raspberry Pi 5

mj22226 opened the pull request:
This PR add a new target "bcm2712" to support the Raspberry Pi 5

Specification:
- Processor Broadcom BCM2712 2.4GHz quad-core 64-bit Arm Cortex-A76 CPU,
- with Cryptographic Extension, 512KB per-core L2 caches, and a
- 2MB shared L3 cache
- Features:
-  VideoCore VII GPU, supporting OpenGL ES 3.1, Vulkan 1.2
-  Dual 4Kp60 HDMI® display output with HDR support
-  4Kp60 HEVC decoder
-  LPDDR4X-4267 SDRAM 4GB and 8GB
-  Dual-band 802.11ac Wi-Fi
-  Bluetooth 5.0 / Bluetooth Low Energy (BLE)
-  microSD card slot, with support for high-speed SDR104 mode
-  2 × USB 3.0 ports, supporting simultaneous 5Gbps operation
-  2 × USB 2.0 ports
-  Gigabit Ethernet, with PoE+ support (requires separate PoE+ HAT)
-  2 × 4-lane MIPI camera/display transceivers
-  PCIe 2.0 x1 interface for fast peripherals (requires separate M.2 HAT or other adapter)
-  5V/5A DC power via USB-C, with Power Delivery support
-  Raspberry Pi standard 40-pin header
-  Real-time clock (RTC), powered from external battery
-  Power button

boot logs: https://gist.github.com/mj22226/5d9c6f1b0b04fa0eacdc493cbcaed6c7
The raspberry pi 5 Wifi chip, same as the rpi 4b, are currently having this issue https://github.com/openwrt/openwrt/issues/13694
a work around at this issue is to replace the firmware with the firmware used on the Raspberry Pi OS https://github.com/RPi-Distro/firmware-nonfree/tree/bullseye/debian/config/brcm80211/cypress
On the runnig system
```
cd /lib/firmware/cypress
wget -O cyfmac43455-sdio.bin https://github.com/RPi-Distro/firmware-nonfree/raw/bullseye/debian/config/brcm80211/cypress/cyfmac43455-sdio-minimal.bin
wget -O cyfmac43455-sdio.clm_blob https://github.com/RPi-Distro/firmware-nonfree/raw/bullseye/debian/config/brcm80211/cypress/cyfmac43455-sdio.clm_blob
cd -
cd /lib/firmware/brcm/
wget -O brcmfmac43455-sdio.raspberrypi,4-model-b.txt https://raw.githubusercontent.com/RPi-Distro/firmware-nonfree/bullseye/debian/config/brcm80211/brcm/brcmfmac43455-sdio.txt
```

mj22226 committed:
linux-firmware: add raspberry pi 5 wifi firmware

Add symbolic link to NVRAM "brcmfmac43455-sdio.raspberrypi,4-model-b.txt"

Signed-off-by: Marty Jones <mj8263788@gmail.com>

hauke commented:
Please mention where you got the kernel patches from in the commit description.
github  thinks that the patch files completely changed, but when manually looking at it I do not see differences in the patches which are looking very similar.

mj22226 commented:
I added a description and left the old patches unchanged, to avoid any confusion.

graysky2 commented:
@mj22226 - wondering if you would default the kernel config for bcm2712 to 16k pagesize?

mj22226 commented:
> @mj22226 - wondering if you would default the kernel config for bcm2712 to 16k pagesize? The bcm2712_defconfig target should give you the needed ksyms.

@graysky2 I initially set the 16KB page size and later changed to 4kb page size for testing, the result is that there are not significant differences.
Testing with geekbench on ubuntu docker container running on OpenWrt

![4k](https://github.com/openwrt/openwrt/assets/67804477/179b1673-5074-49ba-a921-c6d14ceff097)
https://browser.geekbench.com/v6/cpu/3972633

![16k](https://github.com/openwrt/openwrt/assets/67804477/71c55db8-0e95-4be6-b34e-31a8f47f5013)
https://browser.geekbench.com/v6/cpu/3972840

graysky2 commented:
From what I've read, the improvements are related to memory accessing read/writes not to CPU-based tasks.  I would recommend going with 16k frames since that is the default in the `bcm2712_decfonfig`/RPiOS and Arch ARM are shipping their RPi5-specific kernels with this setting.

mj22226 commented:
@graysky2 I going to change it , it doesn't make any differences on OpenWrt.

cyyself commented on target/linux/bcm27xx/bcm2712/config-6.1:
```diff
@@ -0,0 +1,606 @@
+CONFIG_ARM64_ERRATUM_826319=y
+CONFIG_ARM64_ERRATUM_827319=y
+CONFIG_ARM64_ERRATUM_832075=y
+CONFIG_ARM64_ERRATUM_843419=y
```
There are no errata for Cortex-A76. For better system reliability and security, you should at least enable `CONFIG_ARM64_ERRATUM_1418040`, `CONFIG_ARM64_ERRATUM_1165522`, `CONFIG_ARM64_ERRATUM_1286807`, `CONFIG_ARM64_ERRATUM_1463225`. See #14246

mj22226 replied:
Thanks ,  updated.

mj22226 committed:
kernel: add new kernel config symbols

New kernel kernel config symbols found on BCM2712.

Signed-off-by: Marty Jones <mj8263788@gmail.com>

leobsky commented:
Tested on 2 Raspberry Pi 5 4GB and 8GB SDRAM, running stable.
I have connected this https://wiki.geekworm.com/X1000 PCIe extension with a M.2 to PCIe adapter, in order to test a dual 2.5Gb network card with realtek rtl8125 chip.
```
root@openwrt:/#  iperf3 -c 192.168.15.100 -R
Connecting to host 192.168.15.100, port 5201
Reverse mode, remote host 192.168.1.100 is sending
[  5] local 192.168.1.1 port 36086 connected to 192.168.1.100 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec   281 MBytes  2.35 Gbits/sec                  
[  5]   1.00-2.00   sec   280 MBytes  2.35 Gbits/sec                  
[  5]   2.00-3.00   sec   280 MBytes  2.35 Gbits/sec                  
[  5]   3.00-4.00   sec   280 MBytes  2.35 Gbits/sec                  
[  5]   4.00-5.00   sec   280 MBytes  2.35 Gbits/sec                  
[  5]   5.00-6.00   sec   280 MBytes  2.35 Gbits/sec                  
[  5]   6.00-7.00   sec   280 MBytes  2.35 Gbits/sec                  
[  5]   7.00-8.00   sec   280 MBytes  2.35 Gbits/sec                  
[  5]   8.00-9.00   sec   280 MBytes  2.35 Gbits/sec                  
[  5]   9.00-10.00  sec   280 MBytes  2.35 Gbits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-10.00  sec  2.74 GBytes  2.35 Gbits/sec                  sender
[  5]   0.00-10.00  sec  2.74 GBytes  2.35 Gbits/sec                  receiver

iperf Done.
```
```
root@openwrt:/# lspci
0000:00:00.0 PCI bridge: Broadcom Inc. and subsidiaries Device 2712 (rev 21)
0000:01:00.0 PCI bridge: ASMedia Technology Inc. ASM1182e 2-Port PCIe x1 Gen2 Packet Switch
0000:02:03.0 PCI bridge: ASMedia Technology Inc. ASM1182e 2-Port PCIe x1 Gen2 Packet Switch
0000:02:07.0 PCI bridge: ASMedia Technology Inc. ASM1182e 2-Port PCIe x1 Gen2 Packet Switch
0000:03:00.0 Ethernet controller: Realtek Semiconductor Co., Ltd. RTL8125 2.5GbE Controller
0000:04:00.0 Ethernet controller: Realtek Semiconductor Co., Ltd. RTL8125 2.5GbE Controller
0001:00:00.0 PCI bridge: Broadcom Inc. and subsidiaries Device 2712 (rev 21)
0001:01:00.0 Ethernet controller: Device 1de4:0001
```
```
root@openwrt:/# cat /proc/interrupts
           CPU0       CPU1       CPU2       CPU3       
 13:    1015672    1388974    1342354    1536776     GICv2  26 Level     arch_timer
 14:      39537          0          0          0     GICv2  65 Level     107c013880.mailbox
 15:         12          0          0          0     GICv2 153 Level     uart-pl011
 27:          0          0          0          0     GICv2  48 Level     arm-pmu
 28:          0          0          0          0     GICv2  49 Level     arm-pmu
 29:          0          0          0          0     GICv2  50 Level     arm-pmu
 30:          0          0          0          0     GICv2  51 Level     arm-pmu
 36:       2066          0          0          0     GICv2 308 Level     ttyS0
 37:       7236          0          0          0     GICv2 305 Level     mmc0
 38:     141616          0          0          0     GICv2 306 Level     mmc1
 39:          0          0          0          0     GICv2 251 Level     PCIe PME, aerdrv
 44:          0          0          0          0     GICv2 261 Level     PCIe PME, aerdrv
112:          0          0          0          0  rp1_irq_chip   6 Level     eth0
137:          0          0          0          0  rp1_irq_chip  31 Edge      xhci-hcd:usb1
142:          0          0          0          0  rp1_irq_chip  36 Edge      xhci-hcd:usb3
167:    5815364          0          0          0   MIP-MSI 1572864 Edge      eth1
168:    7971046          0          0          0   MIP-MSI 2097152 Edge      eth2
IPI0:      4574       6186       4189       4771       Rescheduling interrupts
IPI1:   1073601    5366598    2447291    2777092       Function call interrupts
IPI2:         0          0          0          0       CPU stop interrupts
IPI3:         0          0          0          0       CPU stop (for crash dump) interrupts
IPI4:         0          0          0          0       Timer broadcast interrupts
IPI5:    257530     365797     313643     408571       IRQ work interrupts
IPI6:         0          0          0          0       CPU wake-up interrupts
Err:          0
```
NVMe M2 SSD test
```
root@openwrt:/# hdparm -tT /dev/nvme0n1

/dev/nvme0n1:
 Timing cached reads:   3442 MB in  2.00 seconds = 1720.73 MB/sec
 Timing buffered disk reads: 2330 MB in  3.00 seconds = 775.75 MB/sec
```

RyanBlakeIT commented:
> Replying to https://github.com/openwrt/openwrt/pull/13987#issuecomment-1872227588

How do you have both an M2 SSD over PCIe and two 2.5Gbps Realtek Ethernet adapters? Are they USB connected or somehow chained to the PCIe? If the latter, how so?

leobsky commented:
> How do you have both an M2 SSD over PCIe and two 2.5Gbps Realtek Ethernet adapters? Are they USB connected or somehow chained to the PCIe? If the latter, how so?

@RyanBlakeIT I tested separately, the connector has a single-lane PCIe.

mllejezebel commented:
> 2.5Gb network card with realtek rtl8125 chip

Would you mind providing a link to the specific dual 2.5Gb network card with the Realtek RTL8125 chip that yielded successful results during testing? I hope to eventually turn my Raspberry Pi 5 into a router. Also thank you @mj22226 for all the work you've done!

ViRb3 commented:
I was testing your build with AWUS036AXML (mt7921u) USB 3.0 WiFi adapter as AP, and the adapter died under a few minutes of sustained load. I had the exact same issue with the official Raspberry Pi OS and tracked down the problem to https://github.com/raspberrypi/linux/issues/5753. A quirk was added to fix it (https://github.com/raspberrypi/linux/pull/5788), and after grabbing that kernel on the Raspberry Pi OS, performance has been flawless for me even after days of stress testing. Sorry if I'm completely wrong, but maybe this quirk needs to be merged in here?

leobsky commented:
@mllejezebel the dual 2.5Gb network card is this https://www.amazon.com/Syba-Gigabit-Ethernet-Express-Controller/dp/B07XJ8CMQX
also tested this card with realtek RTL8111H chip  https://www.amazon.com/Gigabit-Network-RTL8111H-1000Mbps-Ethernet/dp/B0C7RMLGD2
```
root@OpenWrt:~# lspci
0000:00:00.0 PCI bridge: Broadcom Inc. and subsidiaries Device 2712 (rev 21)
0000:01:00.0 PCI bridge: ASMedia Technology Inc. ASM1182e 2-Port PCIe x1 Gen2 Packet Switch
0000:02:03.0 PCI bridge: ASMedia Technology Inc. ASM1182e 2-Port PCIe x1 Gen2 Packet Switch
0000:02:07.0 PCI bridge: ASMedia Technology Inc. ASM1182e 2-Port PCIe x1 Gen2 Packet Switch
0000:03:00.0 Ethernet controller: Realtek Semiconductor Co., Ltd. RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller (rev 15)
0000:04:00.0 Ethernet controller: Realtek Semiconductor Co., Ltd. RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller (rev 15)
0001:00:00.0 PCI bridge: Broadcom Inc. and subsidiaries Device 2712 (rev 21)
0001:01:00.0 Ethernet controller: Device 1de4:0001
```

RPi 5 HAT used on the test:
https://www.amazon.com/Geekworm-X1001-Key-M-Peripheral-Raspberry/dp/B0CPPGGDQT
https://www.amazon.com/Geekworm-X1000-Key-M-Peripheral-Raspberry/dp/B0CQ4D2C9S
with this PCI-E to M.2 adapter https://www.amazon.com/dp/B09C1B2XC3

tltangliang commented:
> @mllejezebel the dual 2.5Gb network card is this https://www.amazon.com/Syba-Gigabit-Ethernet-Express-Controller/dp/B07XJ8CMQX also tested this card with realtek RTL8111H chip https://www.amazon.com/Gigabit-Network-RTL8111H-1000Mbps-Ethernet/dp/B0C7RMLGD2
> 
> ```
> root@OpenWrt:~# lspci
> 0000:00:00.0 PCI bridge: Broadcom Inc. and subsidiaries Device 2712 (rev 21)
> 0000:01:00.0 PCI bridge: ASMedia Technology Inc. ASM1182e 2-Port PCIe x1 Gen2 Packet Switch
> 0000:02:03.0 PCI bridge: ASMedia Technology Inc. ASM1182e 2-Port PCIe x1 Gen2 Packet Switch
> 0000:02:07.0 PCI bridge: ASMedia Technology Inc. ASM1182e 2-Port PCIe x1 Gen2 Packet Switch
> 0000:03:00.0 Ethernet controller: Realtek Semiconductor Co., Ltd. RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller (rev 15)
> 0000:04:00.0 Ethernet controller: Realtek Semiconductor Co., Ltd. RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller (rev 15)
> 0001:00:00.0 PCI bridge: Broadcom Inc. and subsidiaries Device 2712 (rev 21)
> 0001:01:00.0 Ethernet controller: Device 1de4:0001
> ```
> 
> RPi 5 HAT used on the test: https://www.amazon.com/Geekworm-X1001-Key-M-Peripheral-Raspberry/dp/B0CPPGGDQT https://www.amazon.com/Geekworm-X1000-Key-M-Peripheral-Raspberry/dp/B0CQ4D2C9S with this PCI-E to M.2 adapter https://www.amazon.com/dp/B09C1B2XC3

The ASM1182e connects to two 8125 NICs, and when working simultaneously can only provide a combined bandwidth of less than 3Gbps, with substantial speed degradation on both ports.

tltangliang commented:
> Tested on 2 Raspberry Pi 5 4GB and 8GB SDRAM, running stable. I have connected this https://wiki.geekworm.com/X1000 PCIe extension with a M.2 to PCIe adapter, in order to test a dual 2.5Gb network card with realtek rtl8125 chip.
> 
> ```
> root@openwrt:/#  iperf3 -c 192.168.1.100 -R
> Connecting to host 192.168.1.100, port 5201
> Reverse mode, remote host 192.168.1.100 is sending
> [  5] local 192.168.1.1 port 36086 connected to 192.168.1.100 port 5201
> [ ID] Interval           Transfer     Bitrate
> [  5]   0.00-1.00   sec   281 MBytes  2.35 Gbits/sec                  
> [  5]   1.00-2.00   sec   280 MBytes  2.35 Gbits/sec                  
> [  5]   2.00-3.00   sec   280 MBytes  2.35 Gbits/sec                  
> [  5]   3.00-4.00   sec   280 MBytes  2.35 Gbits/sec                  
> [  5]   4.00-5.00   sec   280 MBytes  2.35 Gbits/sec                  
> [  5]   5.00-6.00   sec   280 MBytes  2.35 Gbits/sec                  
> [  5]   6.00-7.00   sec   280 MBytes  2.35 Gbits/sec                  
> [  5]   7.00-8.00   sec   280 MBytes  2.35 Gbits/sec                  
> [  5]   8.00-9.00   sec   280 MBytes  2.35 Gbits/sec                  
> [  5]   9.00-10.00  sec   280 MBytes  2.35 Gbits/sec                  
> - - - - - - - - - - - - - - - - - - - - - - - - -
> [ ID] Interval           Transfer     Bitrate
> [  5]   0.00-10.00  sec  2.74 GBytes  2.35 Gbits/sec                  sender
> [  5]   0.00-10.00  sec  2.74 GBytes  2.35 Gbits/sec                  receiver
> 
> iperf Done.
> ```
> 
> ```
> root@openwrt:/# lspci
> 0000:00:00.0 PCI bridge: Broadcom Inc. and subsidiaries Device 2712 (rev 21)
> 0000:01:00.0 PCI bridge: ASMedia Technology Inc. ASM1182e 2-Port PCIe x1 Gen2 Packet Switch
> 0000:02:03.0 PCI bridge: ASMedia Technology Inc. ASM1182e 2-Port PCIe x1 Gen2 Packet Switch
> 0000:02:07.0 PCI bridge: ASMedia Technology Inc. ASM1182e 2-Port PCIe x1 Gen2 Packet Switch
> 0000:03:00.0 Ethernet controller: Realtek Semiconductor Co., Ltd. RTL8125 2.5GbE Controller
> 0000:04:00.0 Ethernet controller: Realtek Semiconductor Co., Ltd. RTL8125 2.5GbE Controller
> 0001:00:00.0 PCI bridge: Broadcom Inc. and subsidiaries Device 2712 (rev 21)
> 0001:01:00.0 Ethernet controller: Device 1de4:0001
> ```
> 
> ```
> root@openwrt:/# cat /proc/interrupts
>            CPU0       CPU1       CPU2       CPU3       
>  13:    1015672    1388974    1342354    1536776     GICv2  26 Level     arch_timer
>  14:      39537          0          0          0     GICv2  65 Level     107c013880.mailbox
>  15:         12          0          0          0     GICv2 153 Level     uart-pl011
>  27:          0          0          0          0     GICv2  48 Level     arm-pmu
>  28:          0          0          0          0     GICv2  49 Level     arm-pmu
>  29:          0          0          0          0     GICv2  50 Level     arm-pmu
>  30:          0          0          0          0     GICv2  51 Level     arm-pmu
>  36:       2066          0          0          0     GICv2 308 Level     ttyS0
>  37:       7236          0          0          0     GICv2 305 Level     mmc0
>  38:     141616          0          0          0     GICv2 306 Level     mmc1
>  39:          0          0          0          0     GICv2 251 Level     PCIe PME, aerdrv
>  44:          0          0          0          0     GICv2 261 Level     PCIe PME, aerdrv
> 112:          0          0          0          0  rp1_irq_chip   6 Level     eth0
> 137:          0          0          0          0  rp1_irq_chip  31 Edge      xhci-hcd:usb1
> 142:          0          0          0          0  rp1_irq_chip  36 Edge      xhci-hcd:usb3
> 167:    5815364          0          0          0   MIP-MSI 1572864 Edge      eth1
> 168:    7971046          0          0          0   MIP-MSI 2097152 Edge      eth2
> IPI0:      4574       6186       4189       4771       Rescheduling interrupts
> IPI1:   1073601    5366598    2447291    2777092       Function call interrupts
> IPI2:         0          0          0          0       CPU stop interrupts
> IPI3:         0          0          0          0       CPU stop (for crash dump) interrupts
> IPI4:         0          0          0          0       Timer broadcast interrupts
> IPI5:    257530     365797     313643     408571       IRQ work interrupts
> IPI6:         0          0          0          0       CPU wake-up interrupts
> Err:          0
> ```
> 
> NVMe M2 SSD test
> 
> ```
> root@openwrt:/# hdparm -tT /dev/nvme0n1
> 
> /dev/nvme0n1:
>  Timing cached reads:   3442 MB in  2.00 seconds = 1720.73 MB/sec
>  Timing buffered disk reads: 2330 MB in  3.00 seconds = 775.75 MB/sec
> ```

This test is bullshit, both ports have to be tested at the same time for this to work. Refer to the previous test:https://github.com/geerlingguy/raspberry-pi-pcie-devices/issues/46

leobsky commented:
> This test is bullshit, both ports have to be tested at the same time for this to work

@tltangliang no need to be rude when you comment, I just did a basic test, one of the 2.5 ports is connected to WAN and the other 2.5 port to LAN and the card works as expected.

ViRb3 commented:
After some further testing, discovered a few issues. I have 1Gbit symmetrical connection via TPLink UE300.

Enabling irqbalance or packet steering causes the speed to go up to 900Mbit for a second, then quickly drop to and settle at 500Mbit. Disabling both lets me consistently hit 950Mbit.

Also, NAT Loopback seems to be non-functional unless you enable Promiscuous  Mode on the `br-lan` device.

tltangliang commented:
> > This test is bullshit, both ports have to be tested at the same time for this to work
> 
> @tltangliang no need to be rude when you comment, I just did a basic test, one of the 2.5 ports is connected to WAN and the other 2.5 port to LAN and the card works as expected.

@leobsky Your test results are grossly misleading.

Noltari commented:
@mj22226 bcm27xx-userland and bcm27xx-gpu-fw have been updated to the latest version on main branch:
https://github.com/openwrt/openwrt/commit/3df664101a18cf835c97ce5f0fbcc6357a16c101
https://github.com/openwrt/openwrt/commit/e8f55817015112608155a6463ca2d8f5b4ca37b2

You can rebase and remove those two commits :)

Noltari commented:
I've also created a new PR updating bcm27xx-eeprom to the latest version, which adds support for the Raspberry Pi 5:
https://github.com/openwrt/packages/pull/23113

Noltari commented on target/linux/bcm27xx/bcm2712/target.mk:
```diff
@@ -0,0 +1,13 @@
+ARCH:=aarch64
+SUBTARGET:=bcm2712
+BOARDNAME:=BCM2712 boards (64 bit)
+CPU_TYPE:=cortex-a72
```
Shouldn't this be `CPU_TYPE:=cortex-a76`?

mj22226 replied:
I set cortex-a72 because there is not a package feed for cortex-a76, changed now.

Noltari commented on target/linux/bcm27xx/image/config.txt:
```diff
@@ -12,3 +12,8 @@ include distroconfig.txt
 
 [all]
 # Place your custom settings here.
+# usb_max_current_enable=1
+# Enable the PCIe external connector
+# dtparam=pciex1
+# Force Gen 3.0 speeds
+# dtparam=pciex1_gen=3
```
These custom settings are meant for the users, so you shouldn't change this file.
Also, PCIe 3.0 speeds aren't certified and may cause unstability: https://www.raspberrypi.com/documentation/computers/raspberry-pi-5.html#pcie-gen-3-0

> WARNING
> The Raspberry Pi 5 is not certified for Gen 3.0 speeds, and connections to PCIe devices at these speeds may be unstable.

mj22226 replied:
Ok dropped the custom setting

Noltari commented on target/linux/bcm27xx/Makefile:
```diff
@@ -8,8 +8,8 @@ include $(TOPDIR)/rules.mk
 BOARDNAME:=Broadcom BCM27xx
-FEATURES:=audio boot-part display ext4 fpu gpio rootfs-part rtc squashfs usb usbgadget
-SUBTARGETS:=bcm2708 bcm2709 bcm2710 bcm2711
+FEATURES:=audio boot-part display ext4 fpu gpio pci pcie rootfs-part rtc squashfs usb usbgadget
```
PCIe is only present on Raspberry Pi 5, so `pci pcie` should go as `FEATURES+=pci pcie` in bcm2712/target.mk.
Otherwise you're also adding these features for RPi 1/2/3/4.

mj22226 replied:
done.

Noltari commented:
@mj22226 did you try adding Raspberry Pi 5 kernel settings to bcm2709?
I'm not sure if the Raspberry Pi 5 can boot a 32 bit kernel, but the Raspberry Pi 4 can.

Noltari commented on target/linux/bcm27xx/image/distroconfig.txt:
```diff
@@ -12,5 +12,7 @@ dtoverlay=disable-bt
 dtoverlay=disable-bt
 [pi4]
 dtoverlay=disable-bt
+[pi5]
+dtoverlay=disable-bt
```
The Raspberry Pi 5 doesn't have a mini UART, so is this needed? Or was it copied from pi4 without checking?
https://www.raspberrypi.com/documentation/computers/configuration.html#raspberry-pi-5

mj22226 replied:
rpi5 bluetooth work with https://github.com/openwrt/openwrt/pull/10341
```
[   15.939765] Bluetooth: hci0: BCM: features 0x2f
[   15.945700] Bluetooth: hci0: BCM4345C0 Ampak 0073
[   15.950418] Bluetooth: hci0: BCM4345C0 (003.001.025) build 0075
[   15.956856] Bluetooth: hci0: BCM: Using default device address (44:45:c0:00:22:ac)
```

graysky2 commented:
I'm not sure if the Raspberry Pi 5 can boot a 32 bit kernel, but the Raspberry Pi 4 can.

@Noltari - It [cannot](https://forums.raspberrypi.com/viewtopic.php?p=2167386)., "The ARM cores in BCM2712 are incapable of running the kernel in 32-bit mode"

mj22226 committed:
bcm27xx: add new target bcm2712

This patch add support for raspberry pi 5

Specification:
- Processor Broadcom BCM2712 2.4GHz quad-core 64-bit Arm Cortex-A76 CPU,
  with cryptographic extension, 512KB L2 caches per core
Features:
- VideoCore VII GPU, supports OpenGL ES 3.1, Vulkan 1.2
- Dual 4Kp60 HDMI display output with HDR support 4Kp60 HEVC decoder
- LPDDR4X-4267 SDRAM 4GB and 8GB
- Dual-band 802.11ac Wi-Fi
- Bluetooth 5.0 / Bluetooth Low Energy
- microSD card slot, with support for SDR104 high-speed mode
- 2 x USB 3.0 ports
- 2 x USB 2.0 ports
- Gigabit Ethernet
- 2 x 4 lane MIPI camera/display
- PCIe 2.0 x1
- 5V/5A power via USB-C
- Raspberry Pi standard 40-pin header
- Real-time clock RTC
- Power button

Build system: x86_64
Build-tested: bcm2708, bcm2709, bcm2710, bcm2711, bcm2712
Run-tested: bcm2711/RPi4B, bcm2712/RPi5

Signed-off-by: Marty Jones <mj8263788@gmail.com>

mj22226 committed:
bcm27xx: refresh config

Refresh kernel 6.1 config

Signed-off-by: Marty Jones <mj8263788@gmail.com>

mj22226 committed:
bcm27xx: update 6.1 patches to latest version

Add support for BCM2712, Raspberry Pi 5
https://github.com/raspberrypi/linux/commit/3bb5880ab3dd31f75c07c3c33bf29c5d469b28f3
Patches were generated from the diff between linux kernel branch linux-6.1.y
and rpi-6.1.y from raspberry pi kernel source:
- git format-patch linux-6.1.y...rpi-6.1.y

Signed-off-by: Marty Jones <mj8263788@gmail.com>

Noltari commented on target/linux/bcm27xx/bcm2712/config-6.1:
```diff
@@ -0,0 +1,600 @@
+CONFIG_ARCH_WANTS_NO_INSTR=y
+CONFIG_ARCH_WANTS_THP_SWAP=y
+CONFIG_ARM64=y
+CONFIG_ARM64_4K_PAGES=y
```
@mj22226 according to your [previous comment](https://github.com/openwrt/openwrt/pull/13987#issuecomment-1854536319) I thought you were going to change this to 16K pages.
AFAIK, to reduce compatibility issues with certain applications we should go with 4K pages, but I don't think that there's any application used in OpenWrt that isn't compatible with 16K pages (but maybe I'm wrong). And the official Raspberry PI OS is using 16K pages for the RPi 5.
So I'm not really sure on what to do here :).
CC @graysky2

mj22226 replied:
@Noltari I found an issue with 16K pages with squashfs images, when installing multiple app or increasing rootfs size the rpi5 will not boot .

16K pages
```
[    4.942588] mount_root: loading kmods from internal overlay
[    4.954182] kmodloader: loading kernel modules from //etc/modules-boot.d/*
[    4.961263] kmodloader: done loading kernel modules from //etc/modules-boot.d/*
[    5.048328] block: attempting to load /tmp/overlay/upper/etc/config/fstab
[    5.055170] block: unable to load configuration (fstab: Entry not found)
[    5.061923] block: attempting to load /tmp/overlay/etc/config/fstab
[    5.068225] block: unable to load configuration (fstab: Entry not found)
[    5.074954] block: attempting to load /etc/config/fstab
[    5.083796] block: unable to load configuration (fstab: Entry not found)
[    5.090538] block: no usable configuration
[    5.094663] block: attempting to load /etc/config/fstab
[    5.099912] block: unable to load configuration (fstab: Entry not found)
[    5.106639] block: no usable configuration
[    5.110888] loop0: detected capacity change from 0 to 2129920
[    5.158056] loop0: detected capacity change from 2129920 to 2112768
[    5.175107] EXT4-fs (loop0): mounted filesystem with ordered data mode. Quota mode: disabled.
[    5.196700] block: attempting to load /tmp/overlay/upper/etc/config/fstab
[    5.204597] block: extroot: not configured
[    5.208733] block: attempting to load /etc/config/fstab
[    5.213984] block: unable to load configuration (fstab: Entry not found)
[    5.220731] block: no usable configuration
[    5.224981] mount_root: switching to ext4 overlay
[    5.230007] overlayfs: null uuid detected in lower fs '/', falling back to xino=off,index=off,nfs_export=off.
[    5.247626] insmod: module is already loaded - fat
[    5.252952] insmod: module is already loaded - vfat

[    5.046952] mount_root: overlay filesystem in /dev/loop0 has not been formatted yet
[    5.332227] EXT4-fs (loop0): mounted filesystem with ordered data mode. Quota mode: disabled.
[    5.340932] mount_root: overlay filesystem has not been fully initialized yet
[    5.348369] mount_root: switching to ext4 overlay
[    5.357425] overlayfs: null uuid detected in lower fs '/', falling back to xino=off,index=off,nfs_export=off.
[    5.373334] insmod: module is already loaded - fat
[    5.378604] insmod: module is already loaded - vfat
```
4K pages

```
[    6.629601] mount_root: loading kmods from internal overlay
[    6.642790] kmodloader: loading kernel modules from //etc/modules-boot.d/*
[    6.651185] kmodloader: done loading kernel modules from //etc/modules-boot.d/*
[    6.714975] block: attempting to load /tmp/overlay/upper/etc/config/fstab
[    6.721822] block: unable to load configuration (fstab: Entry not found)
[    6.728584] block: attempting to load /tmp/overlay/etc/config/fstab
[    6.734897] block: unable to load configuration (fstab: Entry not found)
[    6.741636] block: attempting to load /etc/config/fstab
[    6.748868] block: unable to load configuration (fstab: Entry not found)
[    6.755612] block: no usable configuration
[    6.759746] block: attempting to load /etc/config/fstab
[    6.765003] block: unable to load configuration (fstab: Entry not found)
[    6.771734] block: no usable configuration
[    6.775987] loop0: detected capacity change from 0 to 2129920
[    6.823466] loop0: detected capacity change from 2129920 to 1864832
[    6.839583] F2FS-fs (loop0): Mounted with checkpoint version = 726ec211
[    6.869192] block: attempting to load /tmp/overlay/upper/etc/config/fstab
[    6.877563] block: extroot: not configured
[    6.881702] block: attempting to load /etc/config/fstab
[    6.887029] block: unable to load configuration (fstab: Entry not found)
[    6.893784] block: no usable configuration
[    6.898574] mount_root: switching to f2fs overlay
[    6.904888] overlayfs: null uuid detected in lower fs '/', falling back to xino=off,index=off,nfs_export=off.
[    6.921569] insmod: module is already loaded - fat
[    6.926943] insmod: module is already loaded - vfat
```
